### PR TITLE
Notification: add action buttons

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
- add action buttons for notifications
- separate notification list item xml
- centralized account -> no need to check in activity for account null
- if action is successful, the item will vanish
- if not successful a snackbar is shown

As @nickvergessen said, currently we use only two buttons.
If more buttons need to be shown, the will shown horizontally and then line wrap.

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

File: src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java [hunk 1]
```diff
@@ -109,21 +204,25 @@ private void downloadIcon(String icon, ImageView itemViewType) {
...
-            activityIcon = (ImageView) itemView.findViewById(R.id.activity_icon);
-            subject = (TextView) itemView.findViewById(R.id.activity_subject);
-            message = (TextView) itemView.findViewById(R.id.activity_message);
-            dateTime = (TextView) itemView.findViewById(R.id.activity_datetime);
+            layout = itemView.findViewById(R.id.notification_layout);
+            icon = itemView.findViewById(R.id.notification_icon);
+            subject = itemView.findViewById(R.id.notification_subject);
+            message = itemView.findViewById(R.id.notification_message);
+            dateTime = itemView.findViewById(R.id.notification_datetime);
+            buttons = itemView.findViewById(R.id.notification_buttons);
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java]: Binding should be done with butterknife, see for example `UserViewHolder`

Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java]: As this was just refactoring, I did not changed it.
But of course it is better to also enhance code in parallel :+1:

**Comment by tobiasKaminsky** [comment]
This looks kind of lost ;-)
So I keep it the old way, maybe you find a nicer way :-)

**Comment by AndyScherzinger** [comment]
Fine with me @tobiasKaminsky 
We can improve this in a later PR 👍

File: src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java [hunk 2]
```diff
@@ -37,57 +41,151 @@
...
+
 /**
  * This Adapter populates a ListView with all notifications for an account within the app.
  */
 public class NotificationListAdapter extends RecyclerView.Adapter<NotificationListAdapter.NotificationViewHolder> {
-    private List<Notification> mValues;
-    private Context context;
+    private static final String TAG = NotificationListAdapter.class.getSimpleName();
+
+    private List<Notification> notifications;
```
Comment (mario) [inline src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java]: notificationsList please.

File: src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java [hunk 3]
```diff
@@ -51,43 +69,120 @@
...
         }
 
+        @Override
+        protected Boolean doInBackground(Action... actions) {
+            HttpMethod method;
+            Action action = actions[0];
+
+            switch (action.type) {
+                case "GET":
+                    method = new GetMethod(action.link);
```
[the remark this answers is not in the document]
Comment (AndyScherzinger) [reply src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java]: My comment was rather about not having the object creation (costs) ;)

**Comment by AndyScherzinger** [comment]
https://github.com/nextcloud/android/pull/2778#issuecomment-402026123 :
https://material.io/design/components/buttons.html#hierarchy-placement --> placement

I'd say let's go for `contained button for high emphasis` (Accept) and a `text button for low emphasis` and better align them because the buttons are to far away from the item's content/text.

@tobiasKaminsky / @mario should we keep it as-in and do this in a follow up PR or should it be changed here right away? I'd actually vote for doing it in a separate PR since this would then also be a design pattern change since we should then have a PR which does this for other screens too, like the file upload chooser. So also hoping for some @nextcloud/designers @jancborchardt feedback :)

**Comment by AndyScherzinger** [comment]
maybe in case we would go for the just text buttons it would be a good solution to right align them so it would basically mimic the layout of pop-up dialogs other than that the question here would be if both should have the same color since atm "decline" is treated like a "cancel action" while I am aware that "decline" or a more destructive action than accept.

**Comment by tobiasKaminsky** [comment]
I would do this in a separate PR.
Then we have plenty of room to discuss this.

--> keep it the way it is here, and discuss in a separate PR: text button or high/med emphasis / change on other places

**Comment by AndyScherzinger** [comment]
:+1: fine with me @tobiasKaminsky

**Comment by AndyScherzinger** [comment]
Needs to be rebased against master now since the link branch has been merged to master

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you have a look here, the count of context leaks is increased by 1 by this PR :/ (thus 1 more lint warning)

**Comment by jancborchardt** [comment]
From the linked Material buttons page, this fits our case nicest:
- Where the outlined button is "Decline" because less important and the filled button is "Accept"
- Both are right-aligned and the primary action is on the right, just like we do in the web interface

**Comment by AndyScherzinger** [comment]
@jancborchardt can you add a separate issue for this? Simply because the button design in @tobiasKaminsky's screenshots fit the design of the buttons throughout the app. SO I'd say this is fine and for visual consistency we should then change it according to your screenshot for all buttons throughout the app but not mix this with the notification change (thus a new issue/pr).

**Comment by jancborchardt** [comment]
Yup! New issue opened at https://github.com/nextcloud/android/issues/2835

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you take care of the lint warning? Context leak?

File: src/main/java/com/owncloud/android/ui/activity/NotificationsActivity.java [hunk 4]
```diff
@@ -221,22 +226,35 @@ public void showFiles(boolean onDeviceOnly) {
...
-        adapter = new NotificationListAdapter(this);
+        try {
+            OwnCloudAccount ocAccount = new OwnCloudAccount(currentAccount, this);
+            client = OwnCloudClientManagerFactory.getDefaultSingleton().getClientFor(ocAccount, this);
+            client.setOwnCloudVersion(AccountUtils.getServerVersion(currentAccount));
+
+            hideRefreshLayoutLoader();
+        } catch (com.owncloud.android.lib.common.accounts.AccountUtils.AccountNotFoundException e) {
+            Log_OC.e(TAG, "Account not found", e);
+        } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/activity/NotificationsActivity.java]: Yes, but different log message, so this is ok.

File: src/main/java/com/owncloud/android/ui/activity/NotificationsActivity.java [hunk 5]
```diff
@@ -221,22 +226,35 @@ public void showFiles(boolean onDeviceOnly) {
...
+            Log_OC.e(TAG, "IO error", e);
+        } catch (OperationCanceledException e) {
```
[the remark this answers is not in the document]
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/activity/NotificationsActivity.java]: Yes, but different log message, so this is ok.

**Comment by tobiasKaminsky** [comment]
https://github.com/nextcloud/android/blob/d19e209ab6d2f91dce183e91af3d93e628f0bcf0/src/main/java/com/owncloud/android/ui/adapter/NotificationListAdapter.java#L142

This is due to static field leak warning.
As mentioned somewhere else (I guess in video streaming PR), we should discuss how to replace this in a proper way and then do it on entire app. Currently this is "our" state of the art for short running background threads.

So I will increase lint and we should tackle this in a separate PR.

**Pull request merged**